File: src/rac/mcp/audit.py
"""Read-access audit recorder for the Guide MCP server (ADR-084).

The deliberate inversion of telemetry (ADR-040): where telemetry is content-free
and silently self-disabling, the audit recorder is content-BEARING by design and
fail-LOUD. It answers the question telemetry cannot — who consulted which
decision, when, and which artifact IDs came back — by appending one JSON line per
MCP read-tool call.

It is default-ABSENT: with no ``audit:`` stanza in ``.rac/config.yaml`` no
recorder is built, no file is created, and the engine's content-free guarantee is
byte-for-byte intact (ADR-084's strict-superset property). When enabled it is
local-only — this module imports no network code (the isolation battery enforces
it); shipping events to a sink is the ``lore-audit`` satellite's job, never the
engine's (ADR-002, ADR-073).

Recording is write-only observability outside the request/response contract
(ADR-032): :func:`observe` returns the tool payload unchanged, so responses are
byte-identical with and without the recorder — except, by design, under
``on_write_error: block`` when a write fails, where the call is refused with a
structured ``audit-unavailable`` error rather than serving un-audited content.

The principal is attributable, not authenticated (ADR-065, ADR-077): it records
who *claimed* to query (git identity by default, ``RAC_AUDIT_PRINCIPAL`` to
override), never a verified identity — the enforced boundary stays repository ACL
plus human pull-request review.
"""

from __future__ import annotations

import json
import os
import secrets
import subprocess
import sys
import time
from collections.abc import Callable
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path

import yaml

from rac.errors import RACError

# Pinned audit event schema version (ADR-084). Bumping it is a recorded decision.
SCHEMA_VERSION = "1"

# Repository config discovery (read-only half of the `.rac/config.yaml` contract).
# Mirrored here rather than imported from ``rac.services.init`` because the MCP
# layer must not import a write-capable service (ADR-031, the isolation battery).
CONFIG_DIR = ".rac"
CONFIG_FILE = "config.yaml"

AUDIT_FILENAME = "audit.jsonl"
PATH_ENV = "RAC_AUDIT_PATH"
PRINCIPAL_ENV = "RAC_AUDIT_PRINCIPAL"
UNATTRIBUTED = "unattributed"
ON_WRITE_ERROR_VALUES = ("warn", "block")
ERROR_AUDIT_UNAVAILABLE = "audit-unavailable"


class MalformedAuditConfig(RACError):
    """The ``audit:`` stanza in ``.rac/config.yaml`` is unreadable (ADR-084).

    A misconfigured audit posture refuses the server start rather than silently
    recording nothing — the compliance control is never quietly off.
    """

    def __init__(self, config_path: str, reason: str) -> None:
        self.config_path = config_path
        super().__init__(f"malformed audit config {config_path}: {reason}")


@dataclass(frozen=True)
class AuditConfig:
    """The ``audit:`` stanza from ``.rac/config.yaml`` (ADR-084)."""

    enabled: bool
    path: str  # the configured path, or "" to use the default
    on_write_error: str  # "warn" | "block"
    config_path: str | None = None


def _find_config_file(start_dir: str) -> Path | None:
    """The nearest ``.rac/config.yaml`` at or above ``start_dir``, or None."""
    current = Path(start_dir).resolve()
    for directory in (current, *current.parents):
        candidate = directory / CONFIG_DIR / CONFIG_FILE
        if candidate.is_file():
            return candidate
    return None


def _state_dir() -> Path:
    base = os.environ.get("XDG_STATE_HOME") or str(Path.home() / ".local" / "state")
    return Path(base) / "rac"


def audit_path(config: AuditConfig | None = None) -> Path:
    """Resolve the audit log path: ``RAC_AUDIT_PATH`` > config ``path`` > XDG default."""
    env_override = os.environ.get(PATH_ENV)
    if env_override:
        return Path(env_override)
    if config is not None and config.path:
        return Path(config.path)
    return _state_dir() / AUDIT_FILENAME


def load_audit_config(start_dir: str) -> AuditConfig:
    """Read the ``audit`` stanza from the nearest ``.rac/config.yaml`` (ADR-084).

    Returns a disabled config when there is no config file or no ``audit``
    section. A malformed shape raises :class:`MalformedAuditConfig` — the audit
    posture is never silently misconfigured.
    """
    disabled = AuditConfig(enabled=False, path="", on_write_error="warn")
    config_path = _find_config_file(start_dir)
    if config_path is None:
        return disabled
    try:
        data = yaml.safe_load(config_path.read_text(encoding="utf-8"))
    except yaml.YAMLError as exc:
        raise MalformedAuditConfig(str(config_path), f"invalid YAML: {exc}") from exc
    section = data.get("audit") if isinstance(data, dict) else None
    if section is None:
        return disabled
    if not isinstance(section, dict):
        raise MalformedAuditConfig(str(config_path), "'audit' must be a mapping")
    enabled = section.get("enabled", False)
    if not isinstance(enabled, bool):
        raise MalformedAuditConfig(str(config_path), "'audit.enabled' must be true or false")
    path = section.get("path", "")
    if not isinstance(path, str):
        raise MalformedAuditConfig(str(config_path), "'audit.path' must be a string")
    on_write_error = section.get("on_write_error", "warn")
    if on_write_error not in ON_WRITE_ERROR_VALUES:
        raise MalformedAuditConfig(
            str(config_path),
            f"'audit.on_write_error' must be one of {', '.join(ON_WRITE_ERROR_VALUES)}",
        )
    return AuditConfig(
        enabled=enabled,
        path=path,
        on_write_error=on_write_error,
        config_path=str(config_path),
    )


def _git_identity(root: str) -> str | None:
    """``"Name <email>"`` from git config in ``root``, or None when unavailable."""

    def _config(key: str) -> str | None:
        try:
            result = subprocess.run(
                ["git", "config", key],
                cwd=root,
                capture_output=True,
                check=False,
                text=True,
            )
        except OSError:  # no git binary, or root is gone
            return None
        if result.returncode != 0:
            return None
        return result.stdout.strip() or None

    name = _config("user.name")
    email = _config("user.email")
    if name and email:
        return f"{name} <{email}>"
    return email or name


def resolve_principal(root: str) -> str:
    """The audit principal: ``RAC_AUDIT_PRINCIPAL`` > git identity > ``unattributed``.

    Attributable, not authenticated (ADR-084): records who *claimed* to query.
    """
    override = os.environ.get(PRINCIPAL_ENV)
    if override and override.strip():
        return override.strip()
    return _git_identity(root) or UNATTRIBUTED


class AuditRecorder:
    """Append-only, content-bearing, fail-loud audit writer (ADR-084).

    Unlike :class:`~rac.mcp.telemetry.TelemetryRecorder`, a write failure is
    never silently swallowed: it is reported once on stderr (fail-loud), and the
    recorder keeps trying, so ``on_write_error: block`` can refuse every
    un-recordable call rather than disabling itself after the first.
    """

    def __init__(self, path: Path, principal: str, on_write_error: str = "warn") -> None:
        self.path = path
        self.principal = principal
        self.on_write_error = on_write_error
        self.session = secrets.token_hex(8)
        self._warned = False

    def record(self, event: dict) -> bool:
        """Append one event line; return True on success, False on write failure."""
        try:
            with open(self.path, "a", encoding="utf-8") as handle:
                handle.write(json.dumps(event, ensure_ascii=False) + "\n")
            return True
        except OSError as exc:
            if not self._warned:
                action = "refusing tool calls" if self.on_write_error == "block" else "continuing"
                print(
                    f"rac mcp: audit write failed ({exc}); {action} "
                    f"(audit.on_write_error={self.on_write_error}, path={self.path}).",
                    file=sys.stderr,
                )
                self._warned = True
            return False


def create_recorder(config: AuditConfig, root: str) -> AuditRecorder | None:
    """Build an audit recorder when enabled, else ``None`` (the default-absent path)."""
    if not config.enabled:
        return None
    path = audit_path(config)
    try:
        path.parent.mkdir(parents=True, exist_ok=True)
    except OSError:
        pass  # the recorder's first write will fail loud
    return AuditRecorder(path, resolve_principal(root), config.on_write_error)


def observe(recorder: AuditRecorder | None, tool: str, args: dict, call: Callable[[], str]) -> str:
    """Run ``call``, record one audit event, and return the payload unchanged.

    With no recorder this is exactly ``call()`` — audit off creates nothing and
    leaves the response byte-identical (ADR-084's strict superset). A raised call
    is recorded as ``outcome: "exception"`` and re-raised, never swallowed
    (ADR-034). Under ``on_write_error: block`` a failed write refuses the call
    with a structured ``audit-unavailable`` error rather than serving un-audited
    content.
    """
    if recorder is None:
        return call()
    started = time.perf_counter()
    try:
        payload = call()
    except BaseException:
        recorder.record(_event(recorder, tool, args, [], "exception", started))
        raise
    event = _event(recorder, tool, args, _returned(payload), _outcome(payload), started)
    if not recorder.record(event) and recorder.on_write_error == "block":
        return json.dumps(
            {"schema_version": SCHEMA_VERSION, "error": ERROR_AUDIT_UNAVAILABLE, "tool": tool},
            ensure_ascii=False,
        )
    return payload


def _event(
    recorder: AuditRecorder,
    tool: str,
    args: dict,
    returned: list[str],
    outcome: str,
    started: float,
) -> dict:
    return {
        "schema_version": SCHEMA_VERSION,
        "ts": datetime.now(UTC).isoformat(timespec="milliseconds").replace("+00:00", "Z"),
        "session": recorder.session,
        "principal": recorder.principal,
        "tool": tool,
        "query": dict(args),
        "returned": returned,
        "outcome": outcome,
        "duration_ms": int((time.perf_counter() - started) * 1000),
    }


def _outcome(payload: str) -> str:
    """``"error"`` when the payload is a structured error, else ``"ok"`` (ADR-034)."""
    try:
        data = json.loads(payload)
    except ValueError:
        return "ok"
    if isinstance(data, dict) and isinstance(data.get("error"), str):
        return "error"
    return "ok"


def _returned(payload: str) -> list[str]:
    """The resolved artifact IDs a call surfaced, parsed from the serialized payload.

    Records IDs only, never bodies (ADR-084): the primary artifact (get_artifact,
    get_related), search/find matches, and get_related incoming + neighborhood
    neighbours. The queried artifact's own outgoing references are raw declared
    target text (``dict[str, list[str]]``), not resolved IDs, so they are not
    recorded as returned access; get_summary surfaces no IDs.
    """
    try:
        data = json.loads(payload)
    except ValueError:
        return []
    if not isinstance(data, dict) or isinstance(data.get("error"), str):
        return []
    ids: list[str] = []
    primary = data.get("id")
    if isinstance(primary, str):
        ids.append(primary)
    for key in ("matches", "incoming", "neighborhood"):
        value = data.get(key)
        if isinstance(value, list):
            ids.extend(
                item["id"]
                for item in value
                if isinstance(item, dict) and isinstance(item.get("id"), str)
            )
    seen: set[str] = set()
    deduped: list[str] = []
    for artifact_id in ids:
        if artifact_id not in seen:
            seen.add(artifact_id)
            deduped.append(artifact_id)
    return deduped

File: rust/decided-mcp/src/audit.rs
//! Read-access audit recorder (ADR-084), ported for the HTTP transport
//! (ADR-098's mandatory-audit-on). Mirrors `src/asdecided/mcp/audit.py`: one JSON
//! line per read-tool call — who claimed to query, which tool, the query args
//! (never content), the artifact IDs returned, the outcome — appended to the
//! configured `.decided/config.yaml` audit sink.
//!
//! Default-absent for stdio (no `audit:` stanza ⇒ no recorder ⇒ byte-unchanged,
//! ADR-084's strict superset). For HTTP the sink is mandatory and the recorder
//! is *shared*: its construction-time principal skips the host git identity
//! (per-call attribution rides `X-AsDecided-Principal`), and a write failure blocks
//! the call rather than serving un-recordable reads.
//!
//! Not a wire-parity surface (the log is a side file with non-deterministic
//! `ts`/`session`/`duration_ms`); the referee compares records field-by-field
//! excluding those three. The line format itself is byte-faithful:
//! `pyjson::dumps_compact` == `json.dumps(event, ensure_ascii=False)`.

use std::path::PathBuf;
use std::time::{Instant, SystemTime, UNIX_EPOCH};

use rac_engine::pyjson::dumps_compact;
use serde_json::{Map, Value};

const SCHEMA_VERSION: &str = "1";
const CONFIG_DIR: &str = ".decided";
const CONFIG_FILE: &str = "config.yaml";
const AUDIT_FILENAME: &str = "audit.jsonl";
const PATH_ENV: &str = "DECIDED_AUDIT_PATH";
const PRINCIPAL_ENV: &str = "DECIDED_AUDIT_PRINCIPAL";
const UNATTRIBUTED: &str = "unattributed";

/// The `audit:` stanza (ADR-084): enabled, configured path, on-write-error mode.
pub struct AuditConfig {
    pub enabled: bool,
    pub path: String,
    pub on_write_error: String, // "warn" | "block"
}

/// Read the `audit` stanza from the nearest `.decided/config.yaml` at or above
/// `root`. No config or no `audit` section ⇒ disabled. Malformed ⇒ Err (the
/// audit posture is never silently misconfigured).
pub fn load_audit_config(root: &str) -> Result<AuditConfig, String> {
    use rac_engine::frontmatter::{yaml_load_config, Yaml};
    let disabled = AuditConfig { enabled: false, path: String::new(), on_write_error: "warn".into() };
    let Some(config_path) = find_config_file(root) else {
        return Ok(disabled);
    };
    let text = std::fs::read_to_string(&config_path).map_err(|e| format!("invalid YAML: {e}"))?;
    let Yaml::Map(pairs) = yaml_load_config(&text).map_err(|p| format!("invalid YAML: {p}"))? else {
        return Ok(disabled);
    };
    let get = |m: &[(Yaml, Yaml)], key: &str| -> Option<Yaml> {
        m.iter().find_map(|(k, v)| match k {
            Yaml::Str(s) if s == key => Some(v.clone()),
            _ => None,
        })
    };
    let Some(section) = get(&pairs, "audit") else {
        return Ok(disabled);
    };
    let Yaml::Map(audit) = section else {
        return Err("'audit' must be a mapping".to_string());
    };
    let enabled = match get(&audit, "enabled") {
        None => false,
        Some(Yaml::Bool(b)) => b,
        Some(_) => return Err("'audit.enabled' must be true or false".to_string()),
    };
    let path = match get(&audit, "path") {
        None => String::new(),
        Some(Yaml::Str(s)) => s,
        Some(_) => return Err("'audit.path' must be a string".to_string()),
    };
    let on_write_error = match get(&audit, "on_write_error") {
        None => "warn".to_string(),
        Some(Yaml::Str(s)) if s == "warn" || s == "block" => s,
        Some(_) => return Err("'audit.on_write_error' must be one of warn, block".to_string()),
    };
    Ok(AuditConfig { enabled, path, on_write_error })
}

/// `DECIDED_AUDIT_PATH` > config `path` > `$XDG_STATE_HOME/decided/audit.jsonl`.
pub fn resolve_audit_path(configured: &str) -> PathBuf {
    if let Some(env) = std::env::var_os(PATH_ENV) {
        if !env.is_empty() {
            return PathBuf::from(env);
        }
    }
    if !configured.is_empty() {
        return PathBuf::from(configured);
    }
    let base = std::env::var_os("XDG_STATE_HOME")
        .map(PathBuf::from)
        .filter(|p| !p.as_os_str().is_empty())
        .unwrap_or_else(|| {
            let home = std::env::var_os("HOME").map(PathBuf::from).unwrap_or_default();
            home.join(".local").join("state")
        });
    base.join("decided").join(AUDIT_FILENAME)
}

pub fn find_config_file(root: &str) -> Option<PathBuf> {
    let start = std::fs::canonicalize(root).unwrap_or_else(|_| PathBuf::from(root));
    std::iter::successors(Some(start.as_path()), |p| p.parent())
        .map(|dir| dir.join(CONFIG_DIR).join(CONFIG_FILE))
        .find(|c| c.is_file())
}

/// The audit principal: `DECIDED_AUDIT_PRINCIPAL` > (git identity, if allowed) >
/// `unattributed`. `allow_git` is false on the shared HTTP server (ADR-098):
/// the checkout's git identity is the host's, never a caller's.
fn resolve_principal(root: &str, allow_git: bool) -> String {
    if let Ok(v) = std::env::var(PRINCIPAL_ENV) {
        if !v.trim().is_empty() {
            return v.trim().to_string();
        }
    }
    if allow_git {
        if let Some(id) = git_identity(root) {
            return id;
        }
    }
    UNATTRIBUTED.to_string()
}

fn git_identity(root: &str) -> Option<String> {
    let cfg = |key: &str| -> Option<String> {
        let out = std::process::Command::new("git")
            .args(["config", key])
            .current_dir(root)
            .output()
            .ok()?;
        if !out.status.success() {
            return None;
        }
        let s = String::from_utf8_lossy(&out.stdout).trim().to_string();
        (!s.is_empty()).then_some(s)
    };
    let name = cfg("user.name");
    let email = cfg("user.email");
    match (name, email) {
        (Some(n), Some(e)) => Some(format!("{n} <{e}>")),
        (n, e) => e.or(n),
    }
}

/// An append-only, content-bearing audit writer (ADR-084).
pub struct Recorder {
    path: PathBuf,
    principal: String,
    on_write_error: String,
    transport: &'static str,
    session: String,
    warned: bool,
}

/// Build a recorder when audit is enabled, else None (default-absent path).
/// `transport` shapes the shared-server posture: on `"http"` the recorder skips
/// the host git identity and blocks on write failure (ADR-098).
pub fn build(root: &str, transport: &'static str, config: &AuditConfig) -> Option<Recorder> {
    if !config.enabled {
        return None;
    }
    let shared = transport == "http";
    let path = resolve_audit_path(&config.path);
    if let Some(parent) = path.parent() {
        let _ = std::fs::create_dir_all(parent); // first write fails loud if not
    }
    Some(Recorder {
        path,
        principal: resolve_principal(root, !shared),
        on_write_error: if shared { "block".to_string() } else { config.on_write_error.clone() },
        transport,
        session: token_hex(8),
        warned: false,
    })
}

impl Recorder {
    fn record(&mut self, event: &Value) -> bool {
        use std::io::Write;
        let line = dumps_compact(event) + "\n";
        match std::fs::OpenOptions::new().create(true).append(true).open(&self.path) {
            Ok(mut f) => f.write_all(line.as_bytes()).is_ok(),
            Err(e) => {
                if !self.warned {
                    let action =
                        if self.on_write_error == "block" { "refusing tool calls" } else { "continuing" };
                    eprintln!(
                        "decided-mcp: audit write failed ({e}); {action} \
(audit.on_write_error={}, path={}).",
                        self.on_write_error,
                        self.path.display()
                    );
                    self.warned = true;
                }
                false
            }
        }
    }
}

/// Run `call`, record one audit event, and return the payload unchanged
/// (ADR-084: audit is observability outside the response contract). With no
/// recorder this is exactly `call()`. Under `on_write_error: block` a failed
/// write refuses the call with a structured `audit-unavailable` error.
pub fn observe(
    recorder: Option<&mut Recorder>,
    request_principal: Option<&str>,
    tool: &str,
    args: Value,
    call: impl FnOnce() -> String,
) -> String {
    let Some(recorder) = recorder else {
        return call();
    };
    let stripped = request_principal.map(str::trim).filter(|s| !s.is_empty());
    let asserted = stripped.is_some();
    let principal = stripped.map(str::to_string).unwrap_or_else(|| recorder.principal.clone());
    let started = Instant::now();
    let payload = call();
    let event = build_event(
        recorder,
        tool,
        args,
        returned_ids(&payload),
        outcome(&payload),
        started,
        &principal,
        asserted,
    );
    if !recorder.record(&event) && recorder.on_write_error == "block" {
        let mut err = Map::new();
        err.insert("schema_version".into(), Value::String(SCHEMA_VERSION.into()));
        err.insert("error".into(), Value::String("audit-unavailable".into()));
        err.insert("tool".into(), Value::String(tool.into()));
        return dumps_compact(&Value::Object(err));
    }
    payload
}

#[allow(clippy::too_many_arguments)]
fn build_event(
    recorder: &Recorder,
    tool: &str,
    args: Value,
    returned: Vec<String>,
    outcome: &str,
    started: Instant,
    principal: &str,
    asserted: bool,
) -> Value {
    let mut e = Map::new();
    e.insert("schema_version".into(), Value::String(SCHEMA_VERSION.into()));
    e.insert("ts".into(), Value::String(iso_now()));
    e.insert("session".into(), Value::String(recorder.session.clone()));
    e.insert("principal".into(), Value::String(principal.into()));
    e.insert("transport".into(), Value::String(recorder.transport.into()));
    e.insert(
        "attribution".into(),
        Value::String(if asserted { "asserted" } else { "local" }.into()),
    );
    e.insert("tool".into(), Value::String(tool.into()));
    e.insert("query".into(), args);
    e.insert("returned".into(), Value::Array(returned.into_iter().map(Value::String).collect()));
    e.insert("outcome".into(), Value::String(outcome.into()));
    e.insert(
        "duration_ms".into(),
        Value::Number((started.elapsed().as_millis() as u64).into()),
    );
    Value::Object(e)
}

/// `"error"` when the payload is a structured error, else `"ok"` (ADR-034).
fn outcome(payload: &str) -> &'static str {
    match serde_json::from_str::<Value>(payload) {
        Ok(Value::Object(m)) if m.get("error").and_then(Value::as_str).is_some() => "error",
        _ => "ok",
    }
}

/// The resolved artifact IDs a call surfaced (IDs only, never bodies, ADR-084):
/// the primary `id`, plus `matches`/`incoming`/`neighborhood` item ids; deduped.
fn returned_ids(payload: &str) -> Vec<String> {
    let Ok(Value::Object(data)) = serde_json::from_str::<Value>(payload) else {
        return Vec::new();
    };
    if data.get("error").and_then(Value::as_str).is_some() {
        return Vec::new();
    }
    let mut ids = Vec::new();
    if let Some(id) = data.get("id").and_then(Value::as_str) {
        ids.push(id.to_string());
    }
    for key in ["matches", "incoming", "neighborhood"] {
        if let Some(Value::Array(items)) = data.get(key) {
            for item in items {
                if let Some(id) = item.get("id").and_then(Value::as_str) {
                    ids.push(id.to_string());
                }
            }
        }
    }
    let mut seen = std::collections::HashSet::new();
    ids.retain(|id| seen.insert(id.clone()));
    ids
}

fn token_hex(nbytes: usize) -> String {
    use std::io::Read;
    let mut buf = vec![0u8; nbytes];
    if std::fs::File::open("/dev/urandom")
        .and_then(|mut f| f.read_exact(&mut buf))
        .is_err()
    {
        // Fallback mix (never byte-refereed); good enough for a session tag.
        let mut state = std::process::id() as u64 ^ 0x9E37_79B9_7F4A_7C15;
        if let Ok(d) = SystemTime::now().duration_since(UNIX_EPOCH) {
            state ^= d.as_nanos() as u64;
        }
        for b in &mut buf {
            state = state.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
            *b = (state >> 33) as u8;
        }
    }
    let mut out = String::with_capacity(nbytes * 2);
    for b in buf {
        use std::fmt::Write as _;
        let _ = write!(out, "{b:02x}");
    }
    out
}

/// `datetime.now(UTC).isoformat(timespec="milliseconds")` with `+00:00`→`Z`.
fn iso_now() -> String {
    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default();
    let secs = now.as_secs() as i64;
    let millis = now.subsec_millis();
    let days = secs.div_euclid(86_400);
    let tod = secs.rem_euclid(86_400);
    let (y, m, d) = civil_from_days(days);
    let (h, mi, s) = (tod / 3600, (tod % 3600) / 60, tod % 60);
    format!("{y:04}-{m:02}-{d:02}T{h:02}:{mi:02}:{s:02}.{millis:03}Z")
}

/// Proleptic-Gregorian civil date from days since 1970-01-01 (Hinnant).
fn civil_from_days(z0: i64) -> (i64, u32, u32) {
    let z = z0 + 719_468;
    let era = if z >= 0 { z } else { z - 146_096 } / 146_097;
    let doe = z - era * 146_097;
    let yoe = (doe - doe / 1460 + doe / 36_524 - doe / 146_096) / 365;
    let y = yoe + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = (doy - (153 * mp + 2) / 5 + 1) as u32;
    let m = if mp < 10 { mp + 3 } else { mp - 9 } as u32;
    (if m <= 2 { y + 1 } else { y }, m, d)
}
